### PR TITLE
Improve selection and merged cell announcements in LibreOffice Calc 7.3 and above

### LINKED ISSUE #6897: LibreOffice Calc: report selected cell coordinates
This is a feature request that NVDA report selected cell  coordinates.
Steps to reproduce:
1. Open Calc.
2. Press shift+downArrow.
NVDA report: nothing
expected: a1 through a2 selected
3. press ctrl+home to move to top.
4. Press shift+space to select entire row
NVDA report: space
expected: a1 through amj1 selected

### LINKED ISSUE #9310: LibreOffice Calc: report merged cell coordinates
### Steps to reproduce:
1. Open Calc.
2. Select a1 through a3.
3. Choose Format - Merge Cells - Merge Cells.
4. Move to cell a1.

#### Actual behavior:
NVDA reports: a1

#### Expected behavior:
NVDA reports: a1 through a3

#### NVDA Installed/portable/running from source:
Installed.

#### NVDA version:
2018.4.1

#### Windows version:
6.1.7601 service pack 1

#### Name and version of other software in use when reproducing the issue:
LibreOffice 6.1.3.2

#### Does the issue still occur after restarting your PC?
Yes.

#### Have you tried any other versions of NVDA?
NO.

### PR DESCRIPTION
**Pull request by LeonarddeR**
### Link to issue number:
Fixes #9310 
Fixes #6897

### Summary of the issue:
In LibreOffice calc, selection is not announced and for merged cells, only the first cell is announced.

### Description of how this pull request fixes the issue:
In LibreOffice 7.3 development branch, support for IAccessibleTable2 and IAccessibleTableCell were added. This allows us to implement something that works as expected.
Daily builds that support this new functionality can be downloaded [here](https://dev-builds.libreoffice.org/daily/master/current.html)

### Testing strategy:
1. Merged several cells and observed that the announcement was appropriate.
2. Selected several cells and observed that the announcement was appropriate.

### Known issues with pull request:
LibreOffice doesn't fire selectionRemove event properly. This is a known issue that's being investigated.

### Change log entries:
Changes
- NVDA will report selection and merged cells in LibreOffice Calc 7.3 and above, to be released in early 2022. (#9310, #6897)

### Code Review Checklist:

- [x] Pull Request description is up to date.
- [x] Unit tests.
- [x] System (end to end) tests.
- [x] Manual testing.
- [x] API is compatible with existing addons.
- [x] User Documentation.
- [x] Change log entry.
- [x] Context sensitive help for GUI changes.
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers

### CODE REVIEW

**Comment by leonardder** [comment]
Thanks to @michaelweghorn who's doing all the great stuff at the Libre side.

File: source/appModules/soffice.py [hunk 1]
```diff
@@ -151,7 +92,9 @@ def _getFormatFieldAndOffsets(self,offset,formatConfig,calculateOffsets=True):
 
 		# optimisation: Assume a hyperlink occupies a full attribute run.
 		try:
-			if obj.IAccessibleTextObject.QueryInterface(IA2.IAccessibleHypertext).hyperlinkIndex(offset) != -1:
+			if obj.IAccessibleTextObject.QueryInterface(
+				IAccessibleHandler.IA2.IAccessibleHypertext
```
Comment (lukaszgo1) [inline source/appModules/soffice.py:96]: Any particular reason why you prefer `IAccessibleHandler.IA2.*` over just `IA2.*`?

Comment (leonardder) [reply source/appModules/soffice.py:96]: Not really, other than that I no longer want to import the typelib stuff from the comInterfaces directory.

**Comment by leonardder** [comment]
I proposed a general way to announce selection in #6959. If people agree, I want to remove selection announcements from this pr and come up with a follow up that deals with selection.

File: source/NVDAObjects/IAccessible/__init__.py [hunk 2]
```diff
@@ -1289,17 +1289,20 @@ def _get_selectionContainer(self):
 			return self.table
 		return super(IAccessible,self).selectionContainer
 
-	def _getSelectedItemsCount_accSelection(self,maxCount):
+	def _getSelectedItemsCount_accSelection(self, maxCount):
```
Comment (seanbudd) [inline source/NVDAObjects/IAccessible/__init__.py:1292]: If this is being changed, could this be updated

```suggestion
	def _getSelectedItemsCount_accSelection(self, maxCount: int) -> int:
```

File: source/NVDAObjects/IAccessible/__init__.py [hunk 3]
```diff
@@ -1310,24 +1313,33 @@ def _getSelectedItemsCount_accSelection(self,maxCount):
...
 
-	def getSelectedItemsCount(self,maxCount):
-		# To fetch the number of selected items, we first try MSAA's accSelection, but if that fails in any way, we fall back to using IAccessibleTable2's nSelectedCells, if we are on an IAccessible2 table.
+	def getSelectedItemsCount(self, maxCount=2):
+		# To fetch the number of selected items, we first try MSAA's accSelection,
+		# but if that fails in any way, we fall back to using IAccessibleTable2's nSelectedCells,
+		# if we are on an IAccessible2 table, or IAccessibleTable's nSelectedChildren,
+		# if we are on an IAccessible table.
 		# Currently Chrome does not implement accSelection, thus for Google Sheets we must use nSelectedCells when on a table.
+		# For older  symphony based products, we use nSelectedChildren.
```
Comment (seanbudd) [inline source/NVDAObjects/IAccessible/__init__.py:1322]:
```suggestion
		# For older symphony based products, we use nSelectedChildren.
```

File: source/appModules/soffice.py [hunk 4]
```diff
@@ -1,82 +1,23 @@
...
-from NVDAObjects.JAB import JAB, JABTextInfo
 from NVDAObjects.IAccessible import IAccessible, IA2TextTextInfo
 from NVDAObjects.behaviors import EditableText
 from logHandler import log
+import speech
+import api
+import braille
+import vision
 
-def gridCoordStringToNumbers(coordString):
```
Comment (seanbudd) [inline source/appModules/soffice.py:20]: Is the removal of these functions and classes going to affect add-ons?

Comment (leonardder) [reply source/appModules/soffice.py:20]: That's very unlikely I think. Add-ons should avoid using appModules directly.

Comment (seanbudd) [reply source/appModules/soffice.py:20]: @feerrenrut @michaelDCurran  what do you think?

**Comment by CyrilleB79** [comment]
@leonardder you wrote:
> That's very unlikely I think. Add-ons should avoid using appModules directly.

I think that this assumption is not correct.

In my [Outlook Extended ](https://github.com/CyrilleB79/outlookExtended)add-on, I import the native Outlook appModule to extend its capability.

The import statements are as follow:
```
# Import all from original outlook appModule to keep it available if needed.
from nvdaBuiltin.appModules.outlook import *
# Import here explecitely used variables from original outlook appModule for clarity.
from nvdaBuiltin.appModules.outlook import UIAGridRow, AddressBookEntry, AppModule
```

**Comment by leonardder** [comment]
Indeed, the only reason why you would like to import an appModule directly is by extending it. With my statement, I meant that it is unlikely that an add-on would import a top level function from an appModule for another reason than extending the appModule. Having said that, I think that, while we should not break NVDA API compatibility other than in Addon API version changing releases, appModules shouldn't be treated as part of the NVDA API. NVDA makes this pretty clear by design, as many objects that are app specific yet reside in the core NVDAObjects package.

**Comment by seanbudd** [comment]
If this PR doesn't preserve API compatibility it will have to wait until 2022.1

**Comment by leonardder** [comment]
I guess that depends on whether the cleanup I did in the soffice appModule for stuff that hasn't been used for years is considered API breaking?

File: source/appModules/soffice.py (hunk 4 above)
Comment (seanbudd) [reply source/appModules/soffice.py:20]: @feerrenrut  - requested you review pending this discussion

**Review by seanbudd — APPROVED** [review]
Thanks for the wait @leonardder

**Pull request merged**